Fix bugs in CMake dependency checking

fmahebert opened the pull request:
## Proposed changes

Fix some bugs revealed while testing out the dependency checking

### Types of changes:

- [x] Bugfix
- [ ] New feature
- [ ] Refactor

### Component:

- [ ] Code
- [ ] Documentation
- [x] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

fmahebert committed:
CMake deps check: Improve tracking of allowed dependencies

fmahebert committed:
CMake deps check: Add memory allocators to allowed targets

fmahebert committed:
CMake deps check: Fix handling of generator expressions

nilsdeppe commented on cmake/SpectreCheckTargetDependencies.cmake:
```diff
@@ -371,6 +371,21 @@ function(_check_and_print_dependencies
+  # and therefore every target has a dependency on B.
+  set(LIST_OF_ALLOWED_EXTRA_TARGETS_AND_DEPS
+    ${LIST_OF_ALLOWED_EXTRA_TARGETS})
+  foreach(_ALLOWED_EXTRA_TARGET ${LIST_OF_ALLOWED_EXTRA_TARGETS})
```
question: wouldn't you want this to be more recursive?As in, SpectreAllocatior depends on Jemalloc, which depends on jemalloc.so, so we would want to add `SpectreAllocator` explicitly, then have this implicitly add `Jemalloc` and `jemalloc.so`

fmahebert replied:
Agreed, I'll fix

nilsdeppe reviewed (commented):
Looks fine, one comment about how this could better handle all nested allowed implicit deps

kidder commented on cmake/SpectreCheckDependencies.cmake:
```diff
@@ -22,6 +22,8 @@ function(check_spectre_libs_dependencies)
       ${SPECTRE_LIBS}
       ALLOWED_EXTRA_TARGETS
       SpectreFlags
+      SpectreAllocator
```
Is order important?  If not, alphabetical?

fmahebert replied:
I've alphabetized. I also removed Jemalloc because the recursive change that Nils suggested made it unnecessary.

fmahebert committed:
Fix incomplete include path

fmahebert commented:
@nilsdeppe, @kidder: I think I've addressed your comments. I squashed the updates directly, because the diffs were nearly as big as the original commits..

nilsdeppe reviewed (approved):
Thank you for doing this :)  LGTM!
